File: src/test/csharp/font/ThousandsSeparator.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using iText.Kernel.Font;
using iText.Layout;
using iText.Layout.Element;
using NUnit.Framework;

namespace iText7.Net_Playground.Font
{
    class ThousandsSeparator
    {
        /// <summary>
        /// Wrong rendering of thousands separator with custom font in itext
        /// https://stackoverflow.com/questions/71322025/wrong-rendering-of-thousands-separator-with-custom-font-in-itext
        /// Brown-Bold.otf 
        /// https://www.swisstransfer.com/d/69d4f4c1-cf2b-4a51-a841-493237b3164f
        /// 
        /// Cannot reproduce the problem. I get "1.000,00 €" for the "fr-BE" culture,
        /// no matter whether it's iText 7.1.17, 7.2.0, or 7.2.1. 
        /// </summary>
        [Test]
        public void testLikeJohnValdevit()
        {
            CultureInfo.CurrentCulture = new CultureInfo("fr-BE");

            // Fonts
            var fontTexte = PdfFontFactory.CreateFont(@"..\..\..\src\test\resources\mkl\testarea\itext7\font\Brown-Bold.otf",
                PdfFontFactory.EmbeddingStrategy.FORCE_EMBEDDED);

            // Initialisation
            var ms = new MemoryStream();

            // Initialize PDF writer
            var writer = new iText.Kernel.Pdf.PdfWriter(ms);
            writer.SetCloseStream(false);

            // Initialize PDF document
            var pdfDoc = new iText.Kernel.Pdf.PdfDocument(writer);

            // Initialize document
            var document = new Document(pdfDoc, iText.Kernel.Geom.PageSize.A4)
                .SetFont(fontTexte).SetFontSize(10);

            // Values
            for (var i = 1000; i <= 2000; i += 100)
                document.Add(new Paragraph(i.ToString("C")));

            // Close document
            document.Close();

            ms.WriteTo(new FileStream(@"..\..\..\target\test-outputs\font\ThousandsSeparatorLikeJohnValdevitDotNet.pdf", FileMode.Create));
        }
    }
}

File: src/test/java/mkl/testarea/itext7/font/ThousandsSeparator.java
package mkl.testarea.itext7.font;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.util.Locale;

import org.junit.BeforeClass;
import org.junit.Test;

import com.itextpdf.io.source.ByteArrayOutputStream;
import com.itextpdf.io.util.StreamUtil;
import com.itextpdf.kernel.font.PdfFont;
import com.itextpdf.kernel.font.PdfFontFactory;
import com.itextpdf.kernel.geom.PageSize;
import com.itextpdf.kernel.pdf.PdfDocument;
import com.itextpdf.kernel.pdf.PdfWriter;
import com.itextpdf.layout.Document;
import com.itextpdf.layout.element.Paragraph;

/**
 * @author mkl
 */
public class ThousandsSeparator {
    final static File RESULT_FOLDER = new File("target/test-outputs", "font");

    @BeforeClass
    public static void setUpBeforeClass() throws Exception {
        RESULT_FOLDER.mkdirs();
    }

    /**
     * <a href="https://stackoverflow.com/questions/71322025/wrong-rendering-of-thousands-separator-with-custom-font-in-itext">
     * Wrong rendering of thousands separator with custom font in itext
     * </a>
     * <br/>
     * <a href="https://www.swisstransfer.com/d/69d4f4c1-cf2b-4a51-a841-493237b3164f">
     * Brown-Bold.otf
     * </a>
     * <p>
     * Cannot reproduce the problem, even in Java. I always get "1.000,00 €" for the "fr-BE" culture.
     * </p>
     */
    @Test
    public void testLikeJohnValdevit() throws IOException {
        try (   InputStream fontResource = getClass().getResourceAsStream("Brown-Bold.otf") ) {
            PdfFont fontTexte = PdfFontFactory.createFont(StreamUtil.inputStreamToArray(fontResource),
                    PdfFontFactory.EmbeddingStrategy.FORCE_EMBEDDED);

            // Initialisation
            ByteArrayOutputStream ms = new ByteArrayOutputStream();

            // Initialize PDF writer
            PdfWriter writer = new PdfWriter(ms);
            writer.setCloseStream(false);

            // Initialize PDF document
            PdfDocument pdfDoc = new PdfDocument(writer);

            // Initialize document
            Document document = new Document(pdfDoc, PageSize.A4).setFont(fontTexte).setFontSize(10);

            // Values
            Locale locale = new Locale("fr", "BE");
            for (int i = 1000; i <= 2000; i += 100)
                document.add(new Paragraph(String.format(locale, "%,d\n", i)));

            // Close document
            document.close();

            Files.write(new File(RESULT_FOLDER, "ThousandsSeparatorLikeJohnValdevit.pdf").toPath(), ms.toByteArray());
        }
    }

}
